Support memory attribution for Model Element

amarchenko opened the pull request:
#### d1807b80b6c27f7300887b4cd0b587931605fa6f
<pre>
Support memory attribution for Model Element
<a href="https://bugs.webkit.org/show_bug.cgi?id=291387">https://bugs.webkit.org/show_bug.cgi?id=291387</a>
<a href="https://rdar.apple.com/133812421">rdar://133812421</a>

Reviewed by Ada Chan and Mike Wyrzykowski.

This commit adds the necessary infrastructure to support memory attribution
in RealityKit within the ModelProcess for the corresponding Web Content Process.
Due to strict permissions the Model Process can not make the calls for
attribution directly, therefore it is originated in the UI Process and
propagated to the GPU Process.
Also, the change depends on availability of task identity token, hence unavailable
on the simulator.

* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/cocoa/GPUProcessCocoa.mm:
(WebKit::GPUProcess::createMemoryAttributionIDForTask):
(WebKit::GPUProcess::unregisterMemoryAttributionID):
Methods to to create and clean up attribution IDs.
GPU Process entitled to do so.

* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp:
(WebKit::ModelConnectionToWebProcess::create):
(WebKit::ModelConnectionToWebProcess::ModelConnectionToWebProcess):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.h:
(WebKit::ModelConnectionToWebProcess::attributionTaskID const):
Stores and tracks attribution ID for the corresponding Web Process.
Has an accessor to the ID for ModelProcessModelPlayerManagerProxy.

* Source/WebKit/ModelProcess/ModelProcess.cpp:
(WebKit::ModelProcess::createModelConnectionToWebProcess):
* Source/WebKit/ModelProcess/ModelProcess.h:
* Source/WebKit/ModelProcess/ModelProcess.messages.in:
Accepts and forwards attribution ID.

* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.cpp:
(WebKit::ModelProcessModelPlayerManagerProxy::createModelPlayer):
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::RKModelLoaderUSD::load):
(WebKit::loadREModelUsingRKUSDLoader):
(WebKit::ModelProcessModelPlayerProxy::create):
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::load):
RKModelLoaderUSD includes attribution ID for load operations.

* Source/WebKit/Scripts/process-entitlements.sh:
Added com.apple.surfboard.shared-simulation-memory-attribution entitlement

* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::createMemoryAttributionIDForTask):
(WebKit::GPUProcessProxy::unregisterMemoryAttributionID):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::createModelProcessConnection):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::disconnectProcess):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::createMemoryAttributionIDIfNeeded):
(WebKit::WebProcessProxy::unregisterMemoryAttributionIDIfNeeded):
* Source/WebKit/UIProcess/WebProcessProxy.h:
Has methods to lazily create and clean up attribution IDs.

* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.load(from:attributionTaskId:completionHandler:)):
(WKSRKEntity.load(from:completionHandler:)): Deleted.
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:
Entity initialization now includes memory attribution option.

Canonical link: <a href="https://commits.webkit.org/293610@main">https://commits.webkit.org/293610@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e7967fcf84f663c88f4c46327185c0b3432c503

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99409 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19059 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9312 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104540 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50010 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19347 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27492 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/75659 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32753 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102416 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14714 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89757 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56018 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/98882 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14509 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49370 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/7827 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106898 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26523 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/84618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26885 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85961 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84132 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21339 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28803 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/6499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/20266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26463 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/31664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26283 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29596 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27850 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (approved):
Thank you, previously discussed / commented offline and looks good, approving

achan00 commented on Source/WebKit/GPUProcess/cocoa/GPUProcessCocoa.mm:
```diff
@@ -168,6 +168,33 @@
+{
+    Ref<WKSharedSimulationConnectionHelper> sharedSimulationConnectionHelper = adoptRef(*new WKSharedSimulationConnectionHelper);
+    sharedSimulationConnectionHelper->unregisterMemoryAttributionID(attributionID, [sharedSimulationConnectionHelper, completionHandler = WTFMove(completionHandler)] (RetainPtr<id> appService) mutable {
+        if (appService.get() != nil)
```
Nit: can just be `if (appService)`

amarchenko replied:
Done.

achan00 commented on Source/WebKit/GPUProcess/cocoa/GPUProcessCocoa.mm:
```diff
@@ -168,6 +168,33 @@
+    Ref<WKSharedSimulationConnectionHelper> sharedSimulationConnectionHelper = adoptRef(*new WKSharedSimulationConnectionHelper);
+    sharedSimulationConnectionHelper->unregisterMemoryAttributionID(attributionID, [sharedSimulationConnectionHelper, completionHandler = WTFMove(completionHandler)] (RetainPtr<id> appService) mutable {
+        if (appService.get() != nil)
+            RELEASE_LOG(ModelElement, "GPUProcess: Memory attribution ID unregistration succeeded");
```
We should log an error if for some reason the unregistration failed.

amarchenko replied:
Done.

achan00 commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1217,6 +1217,44 @@ void WebProcessProxy::modelProcessExited(ProcessTerminationReason reason)
+        return;
+    }
+
+    RELEASE_LOG(ProcessSuspension, "%p - WebProcessProxy is taking a background assertion because a web process is creating a memory attribution ID", this);
```
We should remove this log because we are not actually taking a background assertion here.

amarchenko replied:
Done.

achan00 commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1217,6 +1217,44 @@ void WebProcessProxy::modelProcessExited(ProcessTerminationReason reason)
+    if (!m_memoryAttributionID.has_value())
+        return;
+
+    RELEASE_LOG(ProcessSuspension, "%p - WebProcessProxy is taking a background assertion because a web process is unregistering memory attribution", this);
```
Ditto.

amarchenko replied:
Done.

amarchenko committed:
Support memory attribution for Model Element
https://bugs.webkit.org/show_bug.cgi?id=291387
rdar://133812421

Reviewed by Ada Chan and Mike Wyrzykowski.

This commit adds the necessary infrastructure to support memory attribution
in RealityKit within the ModelProcess for the corresponding Web Content Process.
Due to strict permissions the Model Process can not make the calls for
attribution directly, therefore it is originated in the UI Process and
propagated to the GPU Process.
Also, the change depends on availability of task identity token, hence unavailable
on the simulator.

* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/cocoa/GPUProcessCocoa.mm:
(WebKit::GPUProcess::createMemoryAttributionIDForTask):
(WebKit::GPUProcess::unregisterMemoryAttributionID):
Methods to to create and clean up attribution IDs.
GPU Process entitled to do so.

* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp:
(WebKit::ModelConnectionToWebProcess::create):
(WebKit::ModelConnectionToWebProcess::ModelConnectionToWebProcess):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.h:
(WebKit::ModelConnectionToWebProcess::attributionTaskID const):
Stores and tracks attribution ID for the corresponding Web Process.
Has an accessor to the ID for ModelProcessModelPlayerManagerProxy.

* Source/WebKit/ModelProcess/ModelProcess.cpp:
(WebKit::ModelProcess::createModelConnectionToWebProcess):
* Source/WebKit/ModelProcess/ModelProcess.h:
* Source/WebKit/ModelProcess/ModelProcess.messages.in:
Accepts and forwards attribution ID.

* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.cpp:
(WebKit::ModelProcessModelPlayerManagerProxy::createModelPlayer):
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::RKModelLoaderUSD::load):
(WebKit::loadREModelUsingRKUSDLoader):
(WebKit::ModelProcessModelPlayerProxy::create):
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::load):
RKModelLoaderUSD includes attribution ID for load operations.

* Source/WebKit/Scripts/process-entitlements.sh:
Added com.apple.surfboard.shared-simulation-memory-attribution entitlement

* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::createMemoryAttributionIDForTask):
(WebKit::GPUProcessProxy::unregisterMemoryAttributionID):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::createModelProcessConnection):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::disconnectProcess):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::createMemoryAttributionIDIfNeeded):
(WebKit::WebProcessProxy::unregisterMemoryAttributionIDIfNeeded):
* Source/WebKit/UIProcess/WebProcessProxy.h:
Has methods to lazily create and clean up attribution IDs.

* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.load(from:attributionTaskId:completionHandler:)):
(WKSRKEntity.load(from:completionHandler:)): Deleted.
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:
Entity initialization now includes memory attribution option.

Canonical link: https://commits.webkit.org/293610@main
